### ISSUE TITLE
Bug: error on decode of optional fields on `ConnectProxyConfig`

### ISSUE THREAD

**Issue by abaehreMC** [hashicorp/consul#7418]
#### Overview of the Issue

When sending a service.json file to be registered using the consul HTTP API to the `/v1/agent/register` endpoint any fields that are "optional" cause an error - "Request decode failed: json: unknown field "mesh_gateway"" (obviously the field is variable)

I recreated in playground here - https://play.golang.org/p/tZRpqrV67tC

The `UnmarshalJSON` implementation doesn't care about any fields other than
`DestinationServiceNameSnake`, `DestinationServiceIDSnake`, `LocalServiceAddressSnake`, and `LocalServicePortSnake`.
https://github.com/hashicorp/consul/blob/9e5fd7f925df1066df3d44773a63aee832ac6d24/agent/structs/connect_proxy_config.go#L123

It looks like this is due to the line here https://github.com/hashicorp/consul/pull/6874/files#diff-68348e8d876ea0874ecf1a6390c45bdfR18 which was introduced in 1.7.0.

#### Reproduction Steps

1. Spin up a consul server instance.
2. Hit the `/v1/agent/service/register` endpoint with a service json with fields not listed above, but are valid in the struct in this instance (I ran into the problem with `mesh_gateway`). https://github.com/hashicorp/consul/blob/9e5fd7f925df1066df3d44773a63aee832ac6d24/agent/structs/connect_proxy_config.go#L83

3. Receive a ` Request decode failed: json: unknown field "mesh_gateway"`.

Example json blob:
```
{
  "name": "foo",
  "kind": "connect-proxy",
  "id": "1234",
  "port": 9090,
  "proxy": {
    "destination_service_name": "foo",
    "destination_service_id": "foo-1234",
    "local_service_address": "127.0.0.1",
    "local_service_port": 8443,
    "config": {
      "bind_address": "0.0.0.0",
      "bind_port": "9999",
      "envoy_prometheus_bind_addr": "0.0.0.0:19001",
      "protocol": "http"
    },
    "mesh_gateway": {
      "mode": "local"
    }
  }
}
```

### Operating system and Environment details

This was being run against a docker container running latest (1.7.1 at the time). 
I'm hitting it using the standard go package http lib.

### Additional Notes
As a workaround you can change "mesh_gateway" to "MeshGateway" since that will work with the `Alias` in the UnmarshalJSON.

**Comment by ryansch** [comment]
I just ran into this with the field `"alias_service"` while setting up a sidecar.

**Comment by keefien** [comment]
Is there any Status about that issue?
I'm facing the same behavior by passing "deregister_critical_service_after" in the json, unfortunately the workaround with the UpperCamelCase Alias does not work for me. So is the API complaining about the "id" key in the checks object.

`{
  "ID": "$POD_NAME-metrics",
  "Name": "ceph-exporter",
  "Tags": ["ceph", "monitoring", "metrics"],
  "Address": "$POD_IP",
  "Port":9283,
  "checks": [{"id": "metrics", 
              "name": "HTTP Metrics on port 9283",
              "http": "http://$POD_IP:9283/metrics",
              "interval": "60s",
              "timeout": "10s"
              }],
  "deregister_critical_service_after": "1h"
}`

Can pass the json successfully by removing the mentioned keys, but that doesn't fit my needs.

Used Consul Version: v1.8.4

**Comment by abaehreMC** [comment]
@keefien - I see https://github.com/hashicorp/consul/blob/8beb7164143397915f3487d95d9f810a9102a82c/agent/structs/check_definition.go#L63 which is in the UnmarshalJSON that looks like it fits your json blob. 

Not sure if that is any help or not.

**Comment by keefien** [comment]
@abaehreMC thanks for searching, but it sadly doesn't help. What I should've mentioned before, I've used exactly that configuration since Consul version: v1.5.x the problems first occurred as we jumped to v1.8.x, It's a little hard to understand for me, how this was not recognized during any test on Hashicorp side, as the used option is still in the documentation and not marked as deprecated.

**Comment by rboyer** [comment]
Hi there!

I tracked down a few missing `snake_case` conversions when parsing data into `ConnectProxyConfig` objects. I have fixed them in #8741.

(backstory)
For API calls it's always the case that the `CamelCase` form will work, as that is the canonical format for request fields in the API, and the format that responses will always be in. A subset of the API (mostly related to service registration and discovery) for historical reasons _also_ works with `snake_case` forms. When new fields are added or community PRs merged sometimes the `snake_case` compatibility bridge is overlooked as was done for several fields here.
